src/lib.rs

```diff
@@ -0,0 +1,545 @@
+    let mut state_map = StateMap::new();
+
+    let mut stack = vec![state_group];
+
+    while let Some(prev_state_group) = entry.prev_state_group {
+        stack.push(prev_state_group);
+        if !map.contains_key(&prev_state_group) {
+            panic!("Missing {}", prev_state_group);
+        }
+        entry = &map[&prev_state_group];
+    }
+
+    for sg in stack.iter().rev() {
+        state_map.extend(
+            map[&sg]
+                .state_map
+                .iter()
+                .map(|((t, s), e)| ((t, s), e.clone())),
+        );
+    }
+
+    state_map
+}
+
+// PyO3 INTERFACE STARTS HERE
+
+impl Config {
+    /// Converts string and bool arguments into a Config struct
+    ///
+    /// This function panics if db_url or room_id are empty strings!
+    pub fn new(
+        db_url: String,
+        output_file: String,
+        room_id: String,
+        max_state_group: String,
+        min_saved_rows: String,
+        transactions: bool,
+        level_sizes: String,
+    ) -> Config {
+        if db_url.is_empty() {
+            panic!("db url is required");
+        }
+
+        let mut output: Option<File> = None;
+        if !output_file.is_empty() {
+            output = Some(
+                File::create(output_file)
+                    .unwrap_or_else(|e| panic!("Unable to create output file: {}", e)),
+            );
+        }
+        let output_file = output;
+
+        if room_id.is_empty() {
+            panic!("room_id is required");
+        }
+
+        let mut max_row: Option<i64> = None;
+        if !max_state_group.is_empty() {
+            max_row = Some(
+                max_state_group
+                    .parse()
+                    .expect("max_state_group must be an integer"),
+            );
+        }
+        let max_state_group = max_row;
+
+        let mut min_count: Option<i32> = None;
+        if !min_saved_rows.is_empty() {
+            min_count = Some(
+                min_saved_rows
+                    .parse()
+                    .expect("min_saved_rows must be an integer"),
+            );
+        }
+        let min_saved_rows = min_count;
+
+        let level_sizes: LevelSizes = level_sizes
+            .parse()
+            .unwrap_or_else(|e| panic!("Unable to parse level_sizes: {}", e));
+
+        Config {
+            db_url,
+            output_file,
+            room_id,
+            max_state_group,
+            min_saved_rows,
+            transactions,
+            level_sizes,
+        }
+    }
+}
+
+/// Access point for python code
+///
+/// Default arguments are equivalent to using the command line tool
+/// No default's are provided for db_url or room_id since these arguments
+/// are compulsor (so that new() act's like parse_arguments())
+#[pyfunction(
+    // db_url has no default
+    // room_id  has no default
+    output_file = "String::from(\"\")",
+    max_state_group = "String::from(\"\")",
+    min_saved_rows = "String::from(\"\")",
```

Comment:
I think we should make these args `Optional` and default to `None`

Reply:
done